### PR TITLE
GH-33742: [Python] Address docstrings in Data Types classes 

### LINKED ISSUE #33742: [Python] Address docstrings in Data Types classes
### Describe the enhancement requested

Ensure docstrings for [Data Types Classes](https://arrow.apache.org/docs/python/api/datatypes.html#type-classes) have an Examples section.

### Component(s)

Python

### PR DESCRIPTION
**Pull request by AlenkaF**
### Rationale for this change

Ensure docstrings for [Data Types Classes](https://arrow.apache.org/docs/python/api/datatypes.html#type-classes) have an Examples section.

### What changes are included in this PR?

Docstrings examples are added.

### Are these changes tested?

Yes, with `python -m pytest --doctest-cython python/pyarrow `
* Closes: #33742

### CODE REVIEW

File: python/pyarrow/types.pxi [hunk 1]
```diff
@@ -1009,6 +1471,18 @@ cdef class PyExtensionType(ExtensionType):
...
+    --------
+    Define a UuidType extension type subclassing PyExtensionType:
+
+    >>> import pyarrow as pa
+    >>> class UuidType(pa.PyExtensionType):
+    ...     def __init__(self):
+    ...         pa.PyExtensionType.__init__(self, pa.binary(16))
+    ...     def __reduce__(self):
+    ...         return UuidType, ()
+    ...
```
Comment (AlenkaF) [inline python/pyarrow/types.pxi:1485]: Not sure why, but doctest fails if I proceed with the `PyExtensionType` example and add `>>> uuid_type = UuidType()`.

The failure is:

```python
_______________________________________________________________________________________________________________________ [doctest] pyarrow.lib.PyExtensionType _______________________________________________________________________________________________________________________
EXAMPLE LOCATION UNKNOWN, not showing all tests of that example
??? >>> uuid_type = UuidType()
UNEXPECTED EXCEPTION: PicklingError("Can't pickle <class 'pyarrow.lib.UuidType'>: attribute lookup UuidType on pyarrow.lib failed")
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.9/3.9.16/Frameworks/Python.framework/Versions/3.9/lib/python3.9/doctest.py", line 1334, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest pyarrow.lib.PyExtensionType[2]>", line 1, in <module>
  File "<doctest pyarrow.lib.PyExtensionType[1]>", line 3, in __init__
  File "pyarrow/types.pxi", line 1498, in pyarrow.lib.PyExtensionType.__init__
  File "pyarrow/types.pxi", line 1399, in pyarrow.lib.ExtensionType.__init__
  File "pyarrow/types.pxi", line 1405, in pyarrow.lib.ExtensionType.init
  File "pyarrow/types.pxi", line 1505, in pyarrow.lib.PyExtensionType.__arrow_ext_serialize__
_pickle.PicklingError: Can't pickle <class 'pyarrow.lib.UuidType'>: attribute lookup UuidType on pyarrow.lib failed
/Users/alenkafrim/repos/arrow/python/pyarrow/lib.cpython-39-darwin.so:None: UnexpectedException
```

Will look into it.

**Review by raulcd — COMMENTED** [review]
Ignore this comment. I am testing something on the PR workflow bot.

File: python/pyarrow/types.pxi (hunk 1 above)
Comment (AlenkaF) [reply python/pyarrow/types.pxi:1485]: Ah, seems doctest and pickle do not get along well :D
(related: https://stackoverflow.com/questions/73394835/how-to-test-pickle-ability-in-a-doctest)

Will add a skip here and what is missing to have the example complete.

**Comment by AlenkaF** [comment]
@danepitkin @jorisvandenbossche this is ready for review if you have some time for it.

**Review by raulcd — COMMENTED** [review]
Thanks @AlenkaF , looks great to me. Only a minor nit that I thought might be worth adding but happy to approve if you think is not worth it.

**Review by danepitkin — COMMENTED** [review]
Overall LGTM!

File: python/pyarrow/types.pxi [hunk 2]
```diff
@@ -499,6 +678,13 @@ cdef class StructType(DataType):
             The index of the field with the given name; -1 if the
             name isn't found or there are several fields with the given
             name.
+
+        Examples
+        --------
+        >>> import pyarrow as pa
+        >>> struct_type = pa.struct({'x': pa.int32(), 'y': pa.string()})
+        >>> struct_type.get_field_index('y')
+        1
```
[the remark this answers is not in the document]
Comment (AlenkaF) [reply python/pyarrow/types.pxi:687]: Yeah, I think this is worth adding 👍  Will ping once the code is updated.

**Review by raulcd — APPROVED** [review]
Thanks @AlenkaF ! :+1:

**Pull request merged**